### ISSUE TITLE
kubelet returns 0 for pod cpu values

### ISSUE THREAD

**Issue by chris-sanders** [ubuntu/microk8s#1636]
Please run `microk8s inspect` and attach the generated tarball to this issue.
[inspection-report-20201009_212521.tar.gz](https://github.com/ubuntu/microk8s/files/5357641/inspection-report-20201009_212521.tar.gz)

After enabling the metrics-server I've noticed that node metrics are fine but pod metrics return only memory and always a cpu of 0.

Example of prometheus:
```
kubectl get --raw /apis/metrics.k8s.io/v1beta1/namespaces/lama/pods/prometheus-kps-prometheus-0                                                                                         
{"kind":"PodMetrics","apiVersion":"metrics.k8s.io/v1beta1","metadata":{"name":"prometheus-kps-prometheus-0","namespace":"lama","selfLink":"/apis/metrics.k8s.io/v1beta1/namespaces/lama/pods/prometheus-kps-prometheus-0","creationTimestamp":"2020-10-09T21:13:38Z"},"timestamp":"2020-10-09T21:13:09Z","window":"30s","containers":[{"name":"prometheus-config-reloader","usage":{"cpu":"0","memory":"2628Ki"}},{"name":"rules-configmap-reloader","usage":{"cpu":"0","memory":"2524Ki"}},{"name":"prometheus","usage":{"cpu":"0","memory":"36608Ki"}}]}

```

Looing at all pods in the monitoring space shows it's more than one pods
```
$ kubectl get --raw /apis/metrics.k8s.io/v1beta1/namespaces/lama/pods/ | jq | grep cpu
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
            "cpu": "0",
```

You can check this even without prometheus just using kubectl top which gets it's values from metric-server as well. Here you'll see nodes are fine, pods have no data.

```
$ kubectl top pods -n lama
NAME                                    CPU(cores)   MEMORY(bytes)
alertmanager-kps-alertmanager-0         0m           17Mi
goldilocks-controller-fc9999d96-692z2   0m           12Mi
goldilocks-dashboard-6c5497cccd-j5qtj   0m           13Mi
grafana-7cbcd77bc4-4hr79                0m           87Mi
kps-operator-96f778f4b-c7vs5            0m           21Mi
kube-state-metrics-6f98c6ccc8-md7fk     0m           11Mi
node-exporter-4785r                     0m           9Mi
node-exporter-rb4hh                     0m           10Mi
node-exporter-wshnl                     0m           4Mi
node-exporter-xpnnf                     0m           10Mi
prometheus-kps-prometheus-0             0m           40Mi
vpa-recommender-6b65d9d49f-jczrx        0m           20Mi
vpa-updater-79f7cff5d9-knr47            0m           17Mi

$ kubectl top nodes -n lama
NAME           CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
clever-eel     1794m        7%     32055Mi         66%
enough-husky   9087m        37%    28487Mi         19%
main-cougar    3331m        13%    27902Mi         57%
neat-cub       1494m        6%     28062Mi         58%
```

I've checked and all of the expected flags are enabled on the kube-api-server as per metrics-server documentation:
```
# Enable the aggregation layer
--requestheader-client-ca-file=${SNAP_DATA}/certs/front-proxy-ca.crt
--requestheader-allowed-names=front-proxy-client
--requestheader-extra-headers-prefix=X-Remote-Extra-
--requestheader-group-headers=X-Remote-Group
--requestheader-username-headers=X-Remote-User
--proxy-client-cert-file=${SNAP_DATA}/certs/front-proxy-client.crt
--proxy-client-key-file=${SNAP_DATA}/certs/front-proxy-client.key
#~Enable the aggregation layerr
```

This is on v1.19.2 (1750)

**Comment by chris-sanders** [comment]
Just an update, I chanted the title after a bit more digging, this isn't the metrics server with the issue I think it's kubelet or containerd. Hitting the /stats/summary for kubelet returns 0 for all cpu values, which explains why metrics-server gets all zeros.

**Comment by chris-sanders** [comment]
$ uname -r
5.4.0-48-generic

I've done some more testing locally, this is a recent regression it seems:
1.19/stable 1710: Appears to work fine
1.19/edge   1750: No CPU stats as shown above
latest/edge 1753: Appears to work fine

I don't know what changed in 1750 but that's where the problem is. I'm also not clear where latest/edge comes from that's running 1.19 but ahead of 1.19/edge. Maybe a comparison of these versions can identify a fix for 1750 since I know that's being used for api stability testing to restore cpu metrics.

**Comment by balchua** [comment]
I don't think there is a change in kubelet for containerd between 1.19/edge (1750) and 1.19(1710).
May i know if the pod definitions contain cpu resource requests/limits?

**Comment by chris-sanders** [comment]
I've tried both, I have pods with and with out requests, with and without limits. Neither matters with 1750, both work with 1710 and 1753.

You can reproduce by just installing mk8s, enable metrics server, and `kubectl top -n kube-system`

**Comment by balchua** [comment]
Aaa i was using `latest/edge` thats why i see the pod cpu percentages.
I will try with the `1.19/edge`
Thanks!

**Comment by ktsakalozos** [comment]
@chris-sanders I cannot reproduce this issue. Here are two runs, one with `latest/edge` and one with `1.19/edge`:
```
jackal@aurora:~/workspace/k8s-conformance$ microk8s.kubectl top pods -n kube-system
NAME                                      CPU(cores)   MEMORY(bytes)   
calico-kube-controllers-847c8c99d-2bzgr   3m           9Mi             
calico-node-gkns9                         36m          30Mi            
metrics-server-8bbfb4bdb-z9xcm            1m           13Mi            
jackal@aurora:~/workspace/k8s-conformance$ microk8s.kubectl top nodes -n kube-system
NAME     CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
aurora   1132m        28%    7992Mi          50%       
jackal@aurora:~/workspace/k8s-conformance$ snap list | grep microk8s
microk8s                 v1.19.2                     1750   1.19/edge        canonical*          classic

..................

jackal@aurora:~/workspace/k8s-conformance$ microk8s.kubectl top nodes -n kube-system
NAME     CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
aurora   1171m        29%    8012Mi          50%       
jackal@aurora:~/workspace/k8s-conformance$ microk8s.kubectl top pods -n kube-system
NAME                                      CPU(cores)   MEMORY(bytes)   
calico-kube-controllers-847c8c99d-cb9pd   2m           8Mi             
calico-node-dlbbq                         42m          30Mi            
metrics-server-8bbfb4bdb-k9fdk            1m           12Mi            
jackal@aurora:~/workspace/k8s-conformance$ snap list | grep microk8s
microk8s                 v1.19.2                     1753   latest/edge      canonical*          classic
```
At this point latest and 1.19 build from the master branch. There should not be any differences in them. Would you be able to share the logs (`microk8s insoect` tarball) of the failing setup so we can dig more into the issue?

**Comment by chris-sanders** [comment]
There's a an inspect tar.gz on the first post, I'll spin back up and re-try the /edge locally. I'm on kernel 5.4.0-48-generic both locally and the machines with the issue if that matters. Not sure what kernel you're on.

**Comment by ktsakalozos** [comment]
I am on `Linux aurora 5.4.0-48-generic #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by chris-sanders** [comment]
Well, I'm not seeing it with a local copy of microk8s now either. The cluster I included the original inspection report on is still experiencing this. I don't know what has changed, I'll poke around and see if I can re-produce again locally and I'm happy to grab logs from the cluster that's been running if the initial inspection report gives you any ideas on anything else I should look into on it.

**Comment by chris-sanders** [comment]
Just updating, not yet solved but I've checked the actual cgroups and on the cluster they do not update so I don't think this is kubelet at this point. Running systemd-cgtop shows that while some processes update none of the Kubernetes pods do, with the exception of the calico processes. I'm now looking at my LXD setup, I suspect this has to do with cgroup issues with microk8s + lxd. Not just any LXD, I've already tested that, but probably one of the lxd-in-lxd things I've done.

I'll update here when I find more, I'm pretty convinced now it's not microk8s but I should get this closed out in the next day or two and confirm hopefully.

**Comment by chris-sanders** [comment]
Alright, I can close this issue. It's not a microk8s issue, it's something else with the host cgroups. I have nested containers on the host and one of them has cgroups permissions, something in that nested container is causing the cpu values to not be updated for the containerd containers. I've confirmed by removing all of the nested containers on one node and cpu values are updating as expected.
